cap/doc.go

```diff
@@ -0,0 +1,279 @@
+
+	root
+	|
+	` root/my-worker (spawned goroutine)
+	|
+	` root/my-other-worker (spawned goroutine)
+	|
+	 ` my-subsystem
+		|
+		` <other workers or sub-trees here>
+
+All your program should have allocated the resources and running as expected. If you want
+to stop the system, you can call the Stop method
+
+	stopErr := sup.Stop()
+
+The stopErr error will tell you if any of the children failed to stop. You may
+also join the main goroutine with the root supervisor monitor goroutine using
+the Wait method
+
+	supErr := sup.Wait()
+
+Resources in Workers
+
+There are times where an IO resource belongs to a single worker, and it doesn't
+make sense to consider a worker "started" until this resource is allocated.
+
+For this use case you can create a Worker goroutine that receives a callback
+function that indicates that the Worker was started (or failed to start).
+
+Lets use the NewWorkerWithNotifyStartFn function to model this behavior:
+
+	func newReportGenWorker(
+		name string,
+		dbConfig dbConfig,
+		reportCh chan Report,
+		opts ...cap.WorkerOpt,
+	) cap.Worker {
+		return cap.NewWorkerWithNotifyStartFn(
+			name,
+			func(ctx context.Cotnext, notifyStart cap.NotifyStartFn) error {
+				db, err := openDB(dbConfig)
+				if err != nil {
+					// (1)
+					notifyStart(err)
+					return err
+				}
+				// (2)
+				notifyStart(nil)
+				return runDBWorker(ctx, db)
+			},
+			opts...,
+		)
+	}
+
+	// Look ma, no capataz dependencies
+	func runDBWorker(ctx context.Context, db db.DB, reportCh chan Report) error {
+		// Business Logic Here
+	}
+
+
+We have a Worker that is reponsible to send certain reports to a given channel
+that some consumer is going to use. The Worker goroutine is the only one
+accessing the database, so it is responsible of initializing it. It uses the
+notifyStart function to indicate if there was an error on start when connecting
+to the database. If this happened (1), the start proceduer of the supervision
+system will stop and abort, if not, the Start routine won't continue until the
+worker notifies is ready (2).
+
+What about that channel though?
+
+Where do we create this channel that this producer and possibly a consumer use
+to communicate?
+
+For this, we need to manage resources in sub-trees.
+
+Resources in Sub-trees
+
+For the example above, we could create the channel at the top-level of our
+application (main function) and pass them as arguments there; however, there is
+a drawback doing it that way. If for some reason the channel gets closed, you
+get a worker that is always going to fail, until the error tolerance of all the
+supervisors is met (hopefully) and it crashes hard.
+
+Another approach, is to have the channel getting created in a subtree, and have
+the worker and producer being monitored by it, if for some reason, the
+supervisor fails, it will initialize a new gochan and restart the whole
+operation. To do this, we need to defined a custom BuildNodesFn in our
```

Comment:
Do you mean:

> Another approach is to have a subtree create the channel at the same level as the worker
and producer which the subtree supervisor is monitoring. That way, if for some reason, the
subtree supervisor fails, it will create a new gochan and restart the producer and consumer.